### PR TITLE
[BugFix] Fix RoboHiveEnv tests

### PR DESCRIPTION
**Pull request by sriramsk1999**
## Description

- Make `RoboHiveEnv` tests actually run correctly when executed
- Fixed a bug in testing of envs without cameras  
- Refactor and simplify test case 

## Motivation and Context

Currently, the tests for `RoboHiveEnv` aren't actually executing correctly.  When the env is [instantiated](https://github.com/pytorch/rl/blob/main/test/test_libs.py#L3356), the `from_pixels` key isn't actually used, and none of the envs are tested with `from_pixels=True`.

After the test code is fixed, I get 9 failures for the RoboHive tests, all of which seem to be caused by an issue with `DKitty` environments. 

Short Test log:

```
================================================= short test summary info =================================================
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientFixed-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom_v2d-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandFixed-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom_v2d-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkFixed-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom_v2d-v0-True] - gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed ...
```

Full traceback:

```
-------------------------------------------------- Captured stdout call ---------------------------------------------------
Registered a new env-variant: A:head-A:left-A:right-A:tail-A:tracki-A:tracki-DKittyWalkRandom-v0
________________________________ TestRoboHive.test_robohive[DKittyWalkRandom_v2d-v0-True] _________________________________
Traceback (most recent call last):
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/gym/envs/registration.py", line 120, in spec
    importlib.import_module(mod_name)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'A'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/runner.py", line 342, in from_call
    result: Optional[TResult] = func()
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/runner.py", line 263, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_hooks.py", line 501, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_manager.py", line 119, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 138, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/threadexception.py", line 87, in pytest_runtest_call
    yield from thread_exception_runtest_hook()
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/threadexception.py", line 63, in thread_exception_runtest_hook
    yield
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/unraisableexception.py", line 90, in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/unraisableexception.py", line 65, in unraisable_exception_runtest_hook
    yield
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/logging.py", line 839, in pytest_runtest_call
    yield from self._runtest_for(item, "call")
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/logging.py", line 822, in _runtest_for
    yield
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/capture.py", line 882, in pytest_runtest_call
    return (yield)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 121, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/skipping.py", line 256, in pytest_runtest_call
    return (yield)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 102, in _multicall
    res = hook_impl.function(*args)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/runner.py", line 178, in pytest_runtest_call
    raise e
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/runner.py", line 170, in pytest_runtest_call
    item.runtest()
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/python.py", line 1831, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_hooks.py", line 501, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_manager.py", line 119, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 138, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/pluggy/_callers.py", line 102, in _multicall
    res = hook_impl.function(*args)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/_pytest/python.py", line 194, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/sriramsk/desktop/torchrl/test/test_libs.py", line 3355, in test_robohive
    env = RoboHiveEnv(envname, from_pixels=from_pixels)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/robohive.py", line 54, in __call__
    instance: RoboHiveEnv = super().__call__(*args, **kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/gym.py", line 543, in __call__
    instance: GymWrapper = super().__call__(*args, **kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/common.py", line 166, in __call__
    instance: EnvBase = super().__call__(*args, **kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/gym.py", line 1277, in __init__
    super().__init__(**kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/gym.py", line 731, in __init__
    super().__init__(**kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/common.py", line 2993, in __init__
    self._env = self._build_env(**kwargs)  # writes the self._env attribute
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/robohive.py", line 201, in _build_env
    env = self.lib.make(
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/gym/envs/registration.py", line 156, in make
    return registry.make(id, **kwargs)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/gym/envs/registration.py", line 100, in make
    spec = self.spec(path)
  File "/home/sriramsk/miniconda3/envs/torch_rl/lib/python3.9/site-packages/gym/envs/registration.py", line 123, in spec
    raise error.Error('A module ({}) was specified for the environment but was not found, make sure the package is installed with `pip install` before calling `gym.make()`'.format(mod_name))
gym.error.Error: A module (A) was specified for the environment but was not found, make sure the package is installed with `pip install` before calling `gym.make()`
```

### CODE REVIEW

**Comment by vmoens** [comment]
LGTM
Let's wait until I fix the CI with this
https://github.com/pytorch/rl/pull/2063

**Comment by vmoens** [comment]
The tests are still failing because of missing envs, e.g. DKittyOrientFixed-v0
@vikashplus here is the complete list
```
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientFixed-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientFixed-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom_v2d-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyOrientRandom_v2d-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyOrientRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandFixed-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandFixed-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom_v2d-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyStandRandom_v2d-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyStandRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkFixed-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkFixed-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkFixed-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkRandom-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom_v2d-v0-True] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[DKittyWalkRandom_v2d-v0-False] - gym.error.UnregisteredEnv: No registered env with id: DKittyWalkRandom_v2d-v0
FAILED test/test_libs.py::TestRoboHive::test_robohive[FK1_CloseAllFixed-v4-True] - gym.error.UnregisteredEnv: No registered env with id: FK1_CloseAllFixed-v4
FAILED test/test_libs.py::TestRoboHive::test_robohive[FK1_CloseAllFixed-v4-False] - gym.error.UnregisteredEnv: No registered env with id: FK1_CloseAllFixed-v4
```

**Comment by sriramsk1999** [comment]
Found a fix for the issue! The camera names for the `DKitty` env is causing a problem.

```
['A:trackingZ', 'A:trackingY', 'A:headCam', 'A:tailCam', 'A:leftCam', 'A:rightCam']
```

The colon in the camera name is causing some issue, I'm not sure what that is. But adding `cams = [i.replace("A:", "A_") for i in cams]` fixes the tests. All RoboHive tests are now passing on my machine.

File: torchrl/envs/libs/robohive.py [hunk 1]
```diff
@@ -233,6 +233,7 @@ def register_visual_env(cls, env_name, cams):
 
             if not len(cams):
                 raise RuntimeError("Cannot create a visual envs without cameras.")
+            cams = [i.replace("A:", "A_") for i in cams]
```
Comment (vikashplus) [inline torchrl/envs/libs/robohive.py:236]: Let me know if this is something you think should be changed at RoboHive's end.

Comment (vikashplus) [reply torchrl/envs/libs/robohive.py:236]: Ah, I see your comment above. I believe `A:trackingZ` is a valid string for MuJoCo as well as RoboHive.

Comment (vmoens) [reply torchrl/envs/libs/robohive.py:236]: I would encourage using environment (and derived camera etc) names that are conventional and can be used for serialization. I can imagine someone using the camera name within the experiment name and use the experiment name to save things on disk

**Comment by sriramsk1999** [comment]
Should we merge this PR for now and remove the workaround after the camera names are updated in robohive in the next release? That seems like a reasonable approach.

**Comment by vikashplus** [comment]
I'd vote that we go ahead with the proposed workaround. I'll make sure that we land this change in the next RoboHive release.

File: test/test_libs.py [hunk 2]
```diff
@@ -3338,37 +3338,28 @@ class TestRoboHive:
     # In the CI, robohive should not coexist with other libs so that's fine.
     # Locally these imports can be annoying, especially given the amount of
```
Comment (vikashplus) [inline test/test_libs.py:3339]: RoboHive's prints can be contorled using an env variable
> ROBOHIVE_VERBOSITY=ALL/INFO/(WARN)/ERROR/ONCE/ALWAYS/SILENT

File: torchrl/envs/libs/robohive.py (hunk 1 above)
Comment (vikashplus) [reply torchrl/envs/libs/robohive.py:224]: I have name a note of this for the next release.

**Review by vikashplus — APPROVED** [review]

File: test/test_libs.py (hunk 2 above)
Comment (vmoens) [reply test/test_libs.py:3389]: sure
The goal of this comment is for anyone running the tests and seeing loads of messages from imports. That could be a bit surprising and annoying. We can add that there's a strategy to silence them. I guess one could argue that importing a library should not print anything by default though...

**Comment by sriramsk1999** [comment]
I pulled the latest changes, upgraded robohive to 0.7, and replaced gym with gymnasium. The older failures are fixed, but a new set of tests are now failing.

```
================================================================================================================ short test summary info =================================================================================================================
FAILED test/test_libs.py::TestRoboHive::test_robohive[door-v1-False] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[door-v1-True] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[hammer-v1-False] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[hammer-v1-True] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[pen-v1-False] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[pen-v1-True] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[relocate-v1-False] - ValueError: too many values to unpack (expected 2)
FAILED test/test_libs.py::TestRoboHive::test_robohive[relocate-v1-True] - ValueError: too many values to unpack (expected 2)
=============================================================================================== 8 failed, 1034 passed, 11432 warnings in 921.68s (0:15:21) ===============================================================================================
```

Full traceback:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/robohive.py", line 62, in __call__
    instance._refine_specs()
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/robohive.py", line 303, in _refine_specs
    rollout = self.rollout(2, return_contiguous=False).get("next")
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/common.py", line 2528, in rollout
    tensordict = self.reset()
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/common.py", line 2105, in reset
    tensordict_reset = self._reset(tensordict, **kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/libs/gym.py", line 1165, in _reset
    return super()._reset(tensordict, **kwargs)
  File "/home/sriramsk/desktop/torchrl/torchrl/envs/gym_like.py", line 357, in _reset
    obs, info = self._reset_output_transform(self._env.reset(**kwargs))
ValueError: too many values to unpack (expected 2)
```

**Comment by vmoens** [comment]
Nope it seems that in this case robohive is returning only one (at least torchrl cannot see two things coming out of reset)
The envs are hammer, door, relocate and pen.
This [commit](https://github.com/pytorch/rl/pull/2080/commits/116c7527c8a91acfd9adfc49def59266fcad1574) solves it, and as you can see it does so by providing an empty dict if there is only one output

**Comment by vikashplus** [comment]
[this line](https://github.com/vikashplus/robohive/blob/main/robohive/envs/env_base.py#L513) kinda does the same thing directly in the base class that uses `@implement_for("gymnasium")` to ensure compatibility 
- for the  hammer, door, relocate and pen envs (v1s), I'm overwriting the [reset function](https://github.com/vikashplus/robohive/blob/main/robohive/envs/hands/door_v1.py#L99). Does the `@implement_for("gymnasium")` functionality carry over? 
- Let's say it doesn't,  do you know why it's not affecting other envs ([e.g. 1](https://github.com/vikashplus/robohive/blob/main/robohive/envs/hands/baoding_v1.py#L259)) where we use the same trick for gym compatibility?

**Comment by vmoens** [comment]
Here's my understanding of what is happening: the reset you pointed in env_base does indeed return 2 values but as you were asking just above the new method doesn't and @implement_for doesn't catch that
The way we deal with this in RL is to have one `_process_reset_data` method or similar that we tune to the gym version and never touch. Example
```python
class EnvBase:
    @implement_for("gymnasium")
    def _process_reset_data(self, reset_data):
        if not isinstance(reset_data, tuple) or not (len(reset_data) == 2 and isinstance(reset_data[1], dict)):
            return reset_data, {}
        return reset_data
    @implement_for("gym", None, 0.25) # or whatever other version
    def _process_reset_data(self, reset_data):
        if isinstance(reset_data, tuple) and len(reset_data) == 2 and isinstance(reset_data[1], dict):
            # throw the dict away
            return reset_data[0]
        return reset_data
	def reset(self):
        # your code here
        return self._process_reset_data(reset_data)
class OtherEnv(EnvBase):
	def reset(self):
        # your other code here
        return self._process_reset_data(reset_data)
```
Does that help?

**Comment by vikashplus** [comment]
Thanks @vmoens for digging in. This indeed confirms the theory. 

I was also surprised by why it's not affecting other envs ([e.g. 1](https://github.com/vikashplus/robohive/blob/main/robohive/envs/hands/baoding_v1.py#L259). And there is a clever answer hidden there. Instead of returning the values, these other envs rely on the `super().reset()` for return for which `implement_for ` works perfectly. 
In essence, using `super().reset()` does the same thing as you suggested. I feel it's a bit more elegant, I'll document it and also add a test for it. I'll change that 4 envs that didn't follow this convention and then RoboHive will hopefully take another step toward torchRL compatibility.

**Comment by vmoens** [comment]
`super()` can defo do the job!
We don't use it in RL because that would make the entire Env API dependent on some (nasty) gym stuff + that would make the whole API more complex (people that don't use gym or different versions of gym should not care about that).
But in Robohive's case I guess super() is the most sensible option

**Comment by vikashplus** [comment]
Perfect. Thanks for pairing with me to better understand this issue.

**Pull request merged**